### PR TITLE
Protein serialization

### PR DESCRIPTION
**Pull request by dwhswenson**
This will add protein serialization, which will pass through openff-toolkit but require openff-toolkit 0.11.

Question: Do we combine the `ProteinComponent` and `SmallMoleculeComponent` into a single `MoleculeComponent`? We originally split these because we expected different serialization -- specifically, we planned to write small molecules as SDF and protein as some protein-focused file format (where we never found one that we liked, so we abandoned that approach).

Follow-up: Currently, the `ProteinComponent` has methods to create from PDB or (unimplemented) from PDBx. The `SmallMoleculeComponent` has the ability to get to/from SDF. If combining into a single class, the question is what to do with these methods: allow all (meaning you could read small molecule as PDB -- isn't that problematic?); or remove all. If we remove them, we might still keep conveniences `protein_from_pdb()` and `small_molecule_from_sdf()` as bare methods that return a `MoleculeComponent`.

Note that removing SDF serialization will break network graphml serialization as well, but it is probably better to move the internal representation there to what comes out of our `to_dict`. There's also a case to be made that we no longer have a real need for graphml serialization.

### CODE REVIEW

**Comment by dwhswenson** [comment]
(First commits here are just to get the openff-toolkit 0.11.rc into the environment)

**Comment by dwhswenson** [comment]
Updating status here: Problems are as follows:

1. We don't want to duplicate code for serialization via OpenFF molecule, which we're maintaining in gufe (instead of using OpenFF's internal mechanisms) in order to ensure stability. We can use the same mechanism for proteins and small molecules.
2. Protein serialization will require`openff-toolkit` 0.11. That's currently in RC (although the most recent RC has been out for a month, so hopefully a release comes soon?)

Problem 1 is a design issue on our side. In the original post on this PR, I proposed making a single class for both `ProteinComponent` and `SmallMoleculeComponent`. On reflection, I think the better approach is a single superclass/mixin `ExplicitMoleculeComponent` that handles shared serialization, as well as any other aspects that are common here. This allows the individual classes to add in reasonable on-/off-ramps via PDB, SDF, or whatever other file formats we want to support for that specific molecule "type."

Problem 2 is more a policies issue -- can we allow gufe to depend on the RC, until the code we need is released in final form? Commits in this PR so far are just to get the RC into our CI environment.

**Comment by richardjgowers** [comment]
Re 1 isn’t such a problem and tbh maybe we can merge Small and Protein
components.

Re 2, It’s also that the RC doesn’t have icodes yet anyway. Pinning to the
rc isn’t a drama.

**Comment by RiesBen** [comment]
Things to do:
- [ ] read in pdb_file sucht that a nice rdkit is created
- [ ] fix bond-order
- [ ] fix formal charge 
- [ ] make sure to_pdbfile writes correct file
- [ ] write tests
- [ ] final testing

File: gufe/components/explicitmoleculecomponent.py
Comment (richardjgowers) [inline gufe/components/explicitmoleculecomponent.py:142]: so this looks like it needs the protein "layer" adding in.  One way for this is for `ProteinComponent` to have a `_to_dict` which calls `super()` and then adds in a few extra entries to the dict created here

off the top of my head, extra things include:
- resname, resid, icode per residue
- chain ids per chain
- grouping of atoms into residues (`[[0, 1, 2], [3, 4, 5], ...]`)
- grouping of residues into chains (`[[0, 1, 2, ....], [400, 401, ...]]`)

File: gufe/components/proteincomponent.py
Comment (richardjgowers) [inline gufe/components/proteincomponent.py:110]: insertion codes?

**Comment by richardjgowers** [comment]
also we should try and push the bond order updates to the xml upstream to openmm, then we can hopefully unvendor this...

File: gufe/components/proteincomponent.py
Comment (RiesBen) [reply gufe/components/proteincomponent.py:110]: ? this is for pdb resdiue naming in the write out :)

Comment (RiesBen) [reply gufe/components/proteincomponent.py:110]: I moved it to the writeout - I think that makes it safer.

Comment (RiesBen) [reply gufe/components/proteincomponent.py:110]: ah wait I think I might not know what an insertion code is.

Comment (RiesBen) [reply gufe/components/proteincomponent.py:110]: added it, still don't know what it exactly does ;)

File: gufe/components/explicitmoleculecomponent.py
Comment (RiesBen) [reply gufe/components/explicitmoleculecomponent.py:142]: should be there :)

File: gufe/components/sub_files/data/pdbNames.xml
Comment (RiesBen) [reply gufe/components/sub_files/data/pdbNames.xml:1]: going to happen soon ;)
but its residues.xml at the moment, that changed

Comment (RiesBen) [reply gufe/components/sub_files/data/pdbNames.xml:1]: The current vendored files are only temporary as we plan to push these changes upstream to openMM and then remove these temporary files from our repo again! : )

**Comment by RiesBen** [comment]
Hey everyone,
the merge request should be functional! 
right now waiting for the CI (locally testing was fine).

please ignore formating of any file in sub_files, as they are going to vanish pretty soon again. I will after this merge push these changes to openmm and in a second iteration remove the sub_files folder again.

Thanks for your review!
Cheers

**Comment by RiesBen** [comment]
vendored code  changes were pushed upstream: :)
https://github.com/openmm/openmm/pull/3770

**Comment by RiesBen** [comment]
TODO: One day this should use rdkit's MonomerInfo so that we go into openff correctly. - DONE :)

**Comment by RiesBen** [comment]
Missing pieces:
-[ ] excluding sub_files folder from CI check (help on this pls.)

rest I think is done. (for now)

**Comment by dwhswenson** [comment]
* Can we move `gufe/component/sub_files/` to `gufe/vendor/pdbfile/` or something like that? This way we can block the `vendor` directory from things like coverage, and that will be constant as we move things in and out of vendored status. It's also a more standard way of communicating to code readers that this is vendored code (which means that you shouldn't change it unless necessary).
* I don't think the core repo is the right place for the ipynb file in the dev directory. We need to establish a right place for it (it is a good thing to have), but I don't think we want the churn of notebooks in here if we can avoid.

This is only a partial review, but wanted to get a few thoughts in place ASAP. I'll take a deeper look soon.

**Comment by RiesBen** [comment]
@dwhswenson  good comments, done :)

File: gufe/components/proteincomponent.py
Comment (dwhswenson) [inline gufe/components/proteincomponent.py:514]: This breaks the ability to use `io.StringIO`.

I think you really just want something here like:

```python
if not isinstance(out_path, io.TextIOBase):
    # allows pathlike/str; we close on completion
    out_path = open(out_path, mode='w')
    must_close = True

PDBFile.writeFile(open_top, openmm_pos, file=out_path)

if must_close:
    # we only close the file if we had to open it
    out_path.close()
```

File: gufe/components/proteincomponent.py
Comment (dwhswenson) [inline gufe/components/proteincomponent.py:547]: see comment on pdbfile

File: gufe/components/proteincomponent.py
Comment (dwhswenson) [inline gufe/components/proteincomponent.py:639]:
```suggestion
        raise NotImplementedError()
```

& same throughout

File: gufe/components/proteincomponent.py
Comment (dwhswenson) [inline gufe/components/proteincomponent.py:664]: is it worth even putting the skeletons of these in now? are we even planning to support on-ramp from a .gro of a protein?

File: gufe/components/proteincomponent.py
Comment (dwhswenson) [inline gufe/components/proteincomponent.py:668]: did this not work? It looked like all the code for it should have been implemented.

**Review by dwhswenson — COMMENTED** [review]
Core concepts are fine; lots of little things to change that will help make the code more maintainable in the long run.

My little code samples aren't tested, so I could be wrong in whether they'd work, but the goal is to simplify things where possible.

One other question: Do we want to include all the PDB files or are these files that we can download on the fly? This seems like a lot to add to the repo, and for users to have working tests, it will be part of the footprint of the package as well. @richardjgowers : thoughts on that?

File: gufe/tests/test_proteincomponent.py
Comment (RiesBen) [reply gufe/tests/test_proteincomponent.py:49]: oooooh! ok cool! :)

File: gufe/tests/test_proteincomponent.py
Comment (RiesBen) [reply gufe/tests/test_proteincomponent.py:39]: does this then make sense? 

```
    in_lines = in_lines[1:]
    out_lines = out_lines[1:]
    
    assert len(in_lines) == len(out_lines)
    assert in_lines == out_lines
```

or would you remove the len(x) == len(y) as well? :)

File: gufe/vendor/pdb_file/singelton.py
Comment (RiesBen) [reply gufe/vendor/pdb_file/singelton.py:1]: in my current branch this is correct?  hm... did you change it?

File: gufe/components/proteincomponent.py
Comment (RiesBen) [reply gufe/components/proteincomponent.py:664]: na, I will remove them.

File: gufe/components/proteincomponent.py
Comment (RiesBen) [reply gufe/components/proteincomponent.py]: In theory it should be there, but somehow, there is a problem (maybe with our testfile?)

Comment (RiesBen) [reply gufe/components/proteincomponent.py]: image

something is weird with the connectivity.... as it is 0 for this atom (which it definitly should not be and is correctly assigned in Topology, somehow it is overwritten somewhere?/unlinked?)

File: gufe/components/explicitmoleculecomponent.py
Comment (RiesBen) [reply gufe/components/explicitmoleculecomponent.py:117]: hmmm true, but this would look like this:
image

And if I understand the code correctly the behaviour would change right :)

File: gufe/components/proteincomponent.py
Comment (RiesBen) [reply gufe/components/proteincomponent.py]: Hm I like this behaviour, so the out_path can be created in the function call:
'''python 
out_path = to_pdbFile(out_path=out_dir+"/"+out_file_name)
'''

but I guess depends.

Comment (RiesBen) [reply gufe/components/proteincomponent.py]: might be called lazy

File: gufe/components/proteincomponent.py
Comment (RiesBen) [reply gufe/components/proteincomponent.py]: hm.... ok I got it, the atom names were not recognized by openmm.... fixed.

File: gufe/components/proteincomponent.py
Comment (RiesBen) [reply gufe/components/proteincomponent.py]: hm on TextIOBase I get a mypy error:
image

However I'm not sure what the TextIO type should be, as it seems not to be accesible???

File: gufe/tests/conftest.py
Comment (RiesBen) [reply gufe/tests/conftest.py:84]: I like the idea, but currently I don't manage to implement it.

File: gufe/components/explicitmoleculecomponent.py
Comment (dwhswenson) [reply gufe/components/explicitmoleculecomponent.py:117]: The decorator [`@abc.abstractmethod`](https://docs.python.org/3/library/abc.html#abc.abstractmethod) means that you need to provide an implementation in a subclass -- it will raise an error if you try to create an instance of a subclass that doesn't instantiate this.

That said, it is technically possible to call this as `GufeTokenizable._to_dict(some_obj)`, which would currently return None instead of raising a `NotImplementedError`. I'd say that is incorrect behavior in `GufeTokenizable`, and should be changed there. That one is an edge case where I can't imagine it mattering in practice. However, with `GufeTokenizable._from_dict`, you can actually call `GufeTokenizable._from_dict({})` without error, and you would normally call a classmethod without needing an instance, so this is worth fixing. (Not as part of this PR: I'll fix it separately.) 

However, back to the original point for this PR: you don't need to add these functions here, and it's actually problematic since you don't have them decorated with `abc.abstractmethod`.)

**Review by RiesBen — APPROVED** [review]

**Comment by dwhswenson** [comment]
TODO: (for my own reference when I come back to this on Monday):

- [ ] mark some should-be-impossible sections of code as `-no-cov-`
- [ ] add tests to ensure that all supported input types (string pathname, `pathlib.Path`, `io.StringIO`, `io.TextIOWrapper`) have at least smoke tests to ensure that usage works without error.

This should bring the patch coverage above target (although I might not try to get the patch coverage to 100%).

**Pull request merged**
